### DISCUSSION TITLE
User roles inherit from other roles

### DISCUSSION THREAD

**Discussion by fracosfr** [directus/directus#15347] [Feature Requests]
Hi!

A new suggestion about the users roles :

At the creation of a new user role it will be great if we can select an user role parent.
The rights of the parent will be automatically affected to the new user role and marked as inherited.

So a fourth type of right will be added "**Inherit from parent**" :

- All access
- No access
- Inherit from parent
- Use custom

Then we can force the inherit for the current user role or keep the inherit.

## A little example :
Member role :
![image](https://user-images.githubusercontent.com/92452560/187857279-56c8ec3d-6df8-479c-bdf4-53476326d617.png)

Editor role :
![Screenshot - 01_09_2022 , 09_00_58](https://user-images.githubusercontent.com/92452560/187856364-8a085e7d-0d66-4322-90db-920ec6c8eede.png)

The effective editor role will be :
![image](https://user-images.githubusercontent.com/92452560/187857671-8df1fdd7-ed3e-48c1-8d5d-7d7586e25f17.png)

**Comment by rijkvanzanten** [comment]
Would you ever want to inherit from the parent _and_ set custom permissions at the same time? For example "Allow READ when parent is allowed to read, or when x eq y" 🤔

**Reply by fracosfr** [reply]
Yes the idea is to be able to pass the rights to the children (whether they are "all" or "custom", or "inherited") and then the children could overrite these rights, and these right (overrited or not) can be pass to theirs children.

In an application where there are several user roles, some "custom" rights are the same for all user roles, being able to inherit from a user role would make this type of right more reliable without having to rewrite the right in each user role.

**Reply by rijkvanzanten** [reply]
>  and then the children could overrite these rights, and these right (overrited or not) can be pass to theirs children.

Would it _override_ it, or _add_ to it? For example, if the parent role has "where status = published" and you set the child role to "where status = draft", do you expect the child role to have "where status = draft OR where status = published", or just "where status = draft"?

**Reply by fracosfr** [reply]
Override, when parent has WHERE status="published" the child has WHERE status="published", if the condition need to be different the best way is to change the right to CUSTOM (or ALL ACCESS or NO ACCESS) in the child role.

**Reply by rijkvanzanten** [reply]
Interesting! In that case, you'd lose some of the usefulness of inheriting permissions though right? Cause the moment you want to extend the parent permissions, you have to copy paste all existing permissions 🤔

**Reply by fracosfr** [reply]
No, if the parent is modified, the rights must be automatically propagated (except the forced rights in the children).

**Reply by Michael-Nussbaumer** [reply]
@rijkvanzanten we have a scheduled call about sponsoring this feature on monday with John of your team. We had already a meeting yesterday where we talked about this, and we came to the conclusion, that maybe a little different or additional approach is a better solution.

The new idea was, that a use can have multiple roles which merges the permissions. As in our case, users need different combinations of permissions.

Example:

Collections:
- Brands
- Products
- News
- Sales

1. Users which owns a brand, so can access/edit all brands, the products of the brand, news connected to the brand and sales of products of their brand.
2. Users which only have read permissions for a brand, products but no access to news and sales
3. Users which only have read permissions for a brand, products and news, but no access to sales
4. Users which only have access to the sales/bills of a brand

In this small example you would need a lot of roles to get all possible combinations combined in roles. There are also different users which belong to the same company and don't have access to all the data. For example accounting staff of the brand should only have access to accounting specific data and news for example.

So a solution where you can configure that more modular and add/remove roles from a single user would be way better or a very good addition.

**Reply by Pachat** [reply]
I implemented a simple solution for my company, based on triggers.

- A `role `may have a `parent role` specified in a Directus_roles custom field. If empty, it is a parent role itself.
- On Save (_have here to add a space for instance to `Description` to be able to save_), 
  - if it is a parent role, it propagates any modification to its children with *
  - if it is a child role, it inherits from its parent  with *
  \* exceptions for the fields of collections specified in one of the 4 custom fields added to Directus_roles on `Permissions`, `Validation`, `Presets`, `Fields` which preserve the corresponding fields of the Directus_roles child.

Not as sophisticated as the proposed solutions above but enough for our needs. and easy to maintain.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
